### DISCUSSION TITLE
Allow configuration of cshtml views runtime compilation

### DISCUSSION THREAD

**Discussion by deMD** [umbraco/Umbraco-CMS#16810] [Features and ideas]
Currently by default Umbraco installs uses the InMemory Modelsbuilder and Razor Runtime Compilation. This is fine for installs that use the templating engine within Umbraco and thus is ideal for low code solutions. However, as a software developer, since I started working with Umbraco and ever since V7 I always used the Modelsbuilder SourceCodeAuto mode due to the additional complexity I always needed to add to my application (and who doesn't like that sweet sweet intellisense helping you with your model properties?).

With .Net 6 Microsoft also introduced Hot Reload which is an awesome feature can recompile code (and views) at a certain level dynamically without needing to restart the entire application reducing the need for Razor Runtime Compilation in projects utilizing the Source code ModelsBuilder and an IDE.

Also added in .Net 6 was the possibility to utilize css isolation, a feature that allows you to define css files as children of your views which then get dynamically compiled into css during build (or hot reload). However this feature does not work nicely together with Razor Runtime Compilation.

So here is my suggestion: Add a configuration option to disable Razor Runtime Compilation, with the caveat that you can't use the InMemory Modelsbuilder (akin the Production Runtime Mode) eg:
```json
"Umbraco": {
    "CMS": {
      "Runtime": {
        "DisableRazorRuntimeCompilation": true
      }
    }
  }
  ```
  which would, just like the production runtime mode, prevents this piece of code: https://github.com/umbraco/Umbraco-CMS/blob/contrib/src%2FUmbraco.Web.Common%2FDependencyInjection%2FUmbracoBuilderExtensions.cs#L239-L242. 
  
  Then all that would need to be done manually is to remove this bit of code in the csproj:
```xml
  <PropertyGroup>
        <!-- Remove RazorCompileOnBuild and RazorCompileOnPublish when not using ModelsMode InMemoryAuto -->
        <RazorCompileOnBuild>false</RazorCompileOnBuild>
        <RazorCompileOnPublish>false</RazorCompileOnPublish>
    </PropertyGroup>
```

And add this:
```xml
<ItemGroup>
        <Watch Include="$(MSBuildProjectDirectory)\**\*.cshtml" />
    </ItemGroup>
```

It's interesting that I needed to do this though, as a vanilla .Net MVC project template has neither config in the csproj and `dotnet watch` works fine for cshtml then.

This change would prevent users to have to use the current workaround:
```csharp
public static class CustomUmbracoBuilderExtensions
{
    public static IUmbracoBuilder AddCustomBackOffice(this IUmbracoBuilder builder) => builder
        .AddConfiguration()
        .AddUmbracoCore()
        .AddWebComponents()
        .AddHelpers()
        .AddBackOfficeCore()
        .AddBackOfficeIdentity()
        .AddBackOfficeAuthentication()
        .AddTokenRevocation()
        .AddMembersIdentity()
        .AddUmbracoProfiler()
        .AddMvcAndRazorWithHotReload()
        .AddWebServer()
        .AddRecurringBackgroundJobs()
        .AddNuCache()
        .AddDistributedCache()
        .AddCoreNotifications()
        .AddExamine()
        .AddExamineIndexes();

    private static IUmbracoBuilder AddMvcAndRazorWithHotReload(this IUmbracoBuilder builder, Action<IMvcBuilder>? mvcBuilding = null)
    {
        // TODO: We need to figure out if we can work around this because calling AddControllersWithViews modifies the global app and order is very important
        // this will directly affect developers who need to call that themselves.
        IMvcBuilder mvcBuilder = builder.Services.AddControllersWithViews();

        mvcBuilding?.Invoke(mvcBuilder);

        return builder;
    }
}
```

**Comment by bergmania** [comment]
Hi.. Thanks for the suggestion. It is definitively something we like to look more into in the future.

To my knowledge, we are not allowed to have a dependency to `Microsoft.AspNetCore.Mvc.Razor.RuntimeCompilation` if hot reload should work correctly, so we has to isolate everything related to runtime compilation into it's own assembly which you should then be able to boot Umbraco without. 

I cannot give any timeframe for when we plan do look into it, but we are about to release a options to the dotnet new templates, that can ease the setup of projects as a regular dotnet developer. More info on the PR https://github.com/umbraco/Umbraco-CMS/pull/16815

**Reply by deMD** [reply]
Hmm interesting. As far as  I see in the source code for the RuntimeCompilation package it only disables the watch for .cshtml files, disables build compilation and adds a runtime compiler. With the workaround I showed above and that explicit watch I have noticed no problems with hot reload 🤔

Would it not be possible to just remove that reference (and associated code) when IDEDevelopment is used as development mode?

**Reply by bergmania** [reply]
No we cannot remove assembly dependencies based on configuration
